tools/clang/tools/dxcompiler/dxcshadersourceinfo.cpp

```diff
@@ -0,0 +1,594 @@
+///////////////////////////////////////////////////////////////////////////////
+//                                                                           //
+// dxcshadersourceinfo.cpp                                                   //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Utility helpers for dealing with DXIL part related to shader sources      //
+// and options.                                                              //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxcshadersourceinfo.h"
+#include "dxc/DxilContainer/DxilContainer.h"
+#include "clang/Basic/SourceManager.h"
+#include "clang/Frontend/CodeGenOptions.h"
+#include "llvm/Support/Path.h"
+#include "miniz.h"
+
+using namespace hlsl;
+using Buffer = SourceInfoWriter::Buffer;
+
+static bool ZlibDecompress(const void *pBuffer, size_t BufferSizeInBytes, Buffer *output);
+static bool ZlibCompress(const void *src, size_t srcSize, Buffer *outCompressedData);
+
+
+///////////////////////////////////////////////////////////////////////////////
+// Reader
+///////////////////////////////////////////////////////////////////////////////
+
+llvm::StringRef SourceInfoReader::GetArgs() const {
+  return m_Args;
+}
+llvm::StringRef SourceInfoReader::GetDefines() const {
+  return m_Defines;
+}
+SourceInfoReader::Source SourceInfoReader::GetSource(unsigned i) const {
+  return m_Sources[i];
+}
+unsigned SourceInfoReader::GetSourcesCount() const {
+  return m_Sources.size();
+}
+void SourceInfoReader::Read(const hlsl::DxilSourceInfo *SourceInfo) {
+  const hlsl::DxilSourceInfoSection *section = (const hlsl::DxilSourceInfoSection *)(SourceInfo+1);
```

Comment:
We need a lot more bounds verifications for things here.  Starting with Read accepting a size that came from the blob we are deserializing, then verifying the header fits, then verifying the size in header matches.  And carrying that pattern all the way down into sub-sections.

Reply:
Added a lot of bounds verifications.